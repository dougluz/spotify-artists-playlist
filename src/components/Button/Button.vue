<script lang="ts">
import { defineComponent, PropType } from "@vue/runtime-core"

type ButtonSize = 'sm' | 'md' | 'lg'

type ButtonType = 'primary' | 'secondary' | 'danger'

export default defineComponent({
  name: 'Button',
  props: {
    onClick: {
      required: false,
      default: () => null,
      type: Function as PropType<(...args: any[]) => any>
    },
    label: {
      required: false,
      default: 'Click',
      type: String as PropType<string>
    },
    disabled: {
      required: false,
      default: false,
      type: Boolean as PropType<boolean>
    },
    size: {
      required: false,
      default: 'md',
      type: String as PropType<ButtonSize>
    },
    kind: {
      required: false,
      default: 'primary',
      type: String as PropType<ButtonType>
    }
  }
})
</script>

<template>
  <button :class="size + ' ' + kind" @click="onClick" :disabled="disabled">
    {{label}}
  </button>
</template>

<style lang="sass" scoped>
@use '../../sass/colors'
.button
  color: colors.$text
  border-radius: 500px
  border: 1px solid transparent
  &:hover
    cursor: pointer
  &:disabled
    cursor: not-allowed

.md
  @extend .button
  font-size: 14px
  line-height: 20px
  text-transform: uppercase
  padding: 13px 13px

.lg
  @extend .button
  font-size: 16px
  line-height: 1
  padding: 21px 56px 19px

.primary
  background-color: colors.$primary
  &:active
    background-color: darken(colors.$primary, 10)
  &:disabled
    background-color: darken(colors.$primary, 30)

.secondary
  background-color: transparent
  &:active
    background-color: lighten(colors.$background, 20)
  &:disabled
    color: darken(colors.$text, 30)

.danger
  background-color: colors.$danger
  &:active
    background-color: darken(colors.$danger, 10)
  &:disabled
    background-color: darken(colors.$danger, 30)

</style>