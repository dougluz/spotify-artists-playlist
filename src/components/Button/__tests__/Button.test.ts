import { fireEvent } from '@testing-library/dom'
import { render } from '@testing-library/vue'
import Button from '../Button.vue'

describe('Button tests', () => {
  it('should display the button whitout any props', () => {
    const { container } = render(Button)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the button with the correct label when received through props', async () => {
    const { findByRole } = render(Button, { props: { label: 'Login' } })
    const button = await findByRole('button', { name: /login/i })
    expect(button).toBeInTheDocument()
  })

  it('should call onClick prop when the button is clicked', async () => {
    const onClick = jest.fn()
    const { findByRole } = render(Button, { props: { onClick } })
    const button = await findByRole('button', { name: /click/i })
    fireEvent.click(button)

    expect(onClick).toBeCalledTimes(1)
  })

  it('should be disabled when receiving disabled prop', async () => {
    const { findByRole } = render(Button, { props: { disabled: true } })
    const button = await findByRole('button', { name: /click/i })
    expect(button).toBeDisabled()
  })
})