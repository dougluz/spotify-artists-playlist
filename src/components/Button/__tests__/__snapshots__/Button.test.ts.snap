// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests should display the button whitout any props 1`] = `
<button
  class="md primary"
>
  Click
</button>
`;
